services/QuillLMS/app/services/classlink_integration/user_persister.rb

```diff
@@ -17,27 +17,51 @@ def initialize(incoming_user:, existing_provider_users:, classlink_instance_id:)
     def run
       raise ValidationError, incoming_user.to_s unless self.class.valid?(incoming_user)
 
-      if existing_user
-        updatable_user_hash = self.class.updatable_user(incoming_user)
-        existing_user.user.update!(**updatable_user_hash)
+      if existing_classlink_user
+        update_existing_classlink_user
+      elsif existing_non_classlink_user
+        link_existing_user_to_classlink
       else
-        username = ::GenerateUsername.run(incoming_user[:given_name], incoming_user[:family_name])
-        creatable_user_hash = self.class.creatable_user(incoming_user)
-
-        ActiveRecord::Base.transaction do
-          user = User.new(**creatable_user_hash.merge({ username: }))
-          user.save!(context: ::User::CLASSLINK_USER_VALIDATION_CONTEXT)
-          ClasslinkAccount.create!(
-            external_id: incoming_user[:sourced_id],
-            classlink_instance_id:,
-            user:
-          )
-        end
+        create_new_user_with_classlink
       end
     rescue => e
-      existing_or_new = existing_user ? 'existing' : 'new'
-      msg = "Failed to persist #{existing_or_new} classlink user with sourcedId=#{incoming_user[:sourced_id]}"
-      ErrorNotifier.report(e, { message: msg })
+      handle_error(e)
```

Comment:
Again, this is purely stylistic so take or leave it, but I prefer early return patterns over `if ... else` blocks
```ruby
def run
  return update_existing_classlink_user if existing_classlink_user
  return link_existing_user_to_classroom if existing_non_classlink_user

  create_new_user_with_classlink
rescue => e
  ...
end
```

Reply:
👀